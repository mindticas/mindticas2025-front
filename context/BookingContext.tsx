import React, { createContext, useState, useContext } from 'react';

// Define el tipo de datos que se almacenarÃ¡n en el contexto
type BookingContextType = {
    service: string;
    setService: (service: string) => void;
    dateTime: Date | null;
    setDateTime: (dateTime: Date | null) => void;
    personData: { name: string; lastName: string; phone: string };
    setPersonData: (personData: {
        name: string;
        lastName: string;
        phone: string;
    }) => void;
};

// Crea el contexto
const BookingContext = createContext<BookingContextType | undefined>(undefined);

// Hook personalizado para usar el contexto
export const useBookingContext = () => {
    const context = useContext(BookingContext);
    if (!context) {
        throw new Error(
            'useBookingContext must be used within a BookingProvider',
        );
    }
    return context;
};

// Proveedor del contexto
export const BookingProvider: React.FC<{ children: React.ReactNode }> = ({
    children,
}) => {
    const [service, setService] = useState<string>('');
    const [dateTime, setDateTime] = useState<Date | null>(null);
    const [personData, setPersonData] = useState<{
        name: string;
        lastName: string;
        phone: string;
    }>({
        name: '',
        lastName: '',
        phone: '',
    });

    return (
        <BookingContext.Provider
            value={{
                service,
                setService,
                dateTime,
                setDateTime,
                personData,
                setPersonData,
            }}
        >
            {children}
        </BookingContext.Provider>
    );
};
