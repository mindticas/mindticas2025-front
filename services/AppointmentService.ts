import { AppointmentCreate } from "@/interfaces/appointment/AppointmentCreate";
import { API_URL } from "./apiConfig";
import { Appointment } from "@/interfaces/appointment/Appointment";

export const createAppointment = async(newAppointment: AppointmentCreate): Promise<Appointment> => {
    try {
        const response = await fetch(`${API_URL}/appointment`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(newAppointment),
        });
        if (!response.ok) {
            throw new Error('Error to create appointment');
        }
        return await response.json() 
}   catch(error){
    throw error;
}}

export const getAppointments = async(): Promise<Appointment[]> => {
    try {
        const response = await fetch(`${API_URL}/appointment`);
        if (!response.ok) {
            throw new Error('Error to get appointments');
        }
        return await response.json() 
}   catch(error){
    throw error;
}}