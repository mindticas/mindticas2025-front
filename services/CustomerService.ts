import { Customer } from "@/interfaces/customer/Customer";
import { API_URL } from "./apiConfig";
import { SortCriteria } from "@/components/SearchFilters";

export const getCustomers = async (sort?: SortCriteria): Promise<Customer[]> => {
    try {
        const url = new URL(`${API_URL}/customer`);
        if (sort) {
            url.searchParams.append('param', sort);
        }

        const response = await fetch(url.toString());
        
        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.message || 'Error al obtener los clientes');
        }
        
        return await response.json();
    } catch (error) {
        throw error;
    }
}

export const getCustomerById = async (customerId: number): Promise<Customer> => {
    try {
        const response = await fetch(`${API_URL}/customer/${customerId}`);
        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.message || 'Error al obtener el cliente');
        }
        return await response.json();
    } catch (error) {
        throw error;
    }
}