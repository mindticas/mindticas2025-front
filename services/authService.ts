import { LoginCredentials } from '@/interfaces/login/LoginCredentials';
import { LoginResponse } from '@/interfaces/login/LoginResponse';
import { API_URL } from './apiConfig';
import Cookies from 'js-cookie';
import router from 'next/router';

export const loginUser = async (
    credentials: LoginCredentials,
): Promise<LoginResponse> => {
    try {
        const response = await fetch(`${API_URL}/auth/login`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(credentials),
        });
        if (!response.ok) {
            throw new Error(
                'Usuario o contraseÃ±a incorrectos, intente nuevamente',
            );
        }
        const userData: LoginResponse = await response.json();
        const { token } = userData;

        // Save cookie
        Cookies.set('AUTH_TOKEN', token, {
            expires: 90,
            sameSite: 'strict',
        });
        return userData;
    } catch (error) {
        console.error('Error en el login');
        throw error;
    }
};

export const handleLogout = () => {
    try {
        Cookies.remove('AUTH_TOKEN');
        router.push('/login');
    } catch (error) {
        console.error('Error during logout:', error);
    }
};
