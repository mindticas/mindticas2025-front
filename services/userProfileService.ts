import { UserProfile } from '@/interfaces/userProfile/userProfile';
import { API_URL } from './apiConfig';

import Cookies from 'js-cookie';

export const getUserProfile = async (): Promise<UserProfile[]> => {
    try {
        const response = await fetch(`${API_URL}/userProfile`, {
            method: 'GET',
            headers: { 'Content-Type': 'application/json' },
        });

        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Error ${response.status}: ${error}`);
        }
        return await response.json();
    } catch (error) {
        throw error;
    }
};

export const updateUserProfile = async (
    data: Partial<UserProfile>,
): Promise<UserProfile> => {
    try {
        const token = Cookies.get('AUTH_TOKEN');
        const response = await fetch(`${API_URL}/userProfile/1`, {
            method: 'PATCH',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Bearer ${token}`,
            },
            body: JSON.stringify(data),
        });

        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Error ${response.status}: ${error}`);
        }
        return await response.json();
    } catch (error) {
        throw error;
    }
};
