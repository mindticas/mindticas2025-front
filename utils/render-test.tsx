import { Provider } from '@/components/ui/provider';
import { BookingProvider } from '@/context/BookingContext';
import { render as rtlRender } from '@testing-library/react';

export function render(ui: React.ReactNode) {
    return rtlRender(<>{ui}</>, {
        wrapper: (props: React.PropsWithChildren) => (
            <Provider>
                <BookingProvider>{props.children}</BookingProvider>
            </Provider>
        ),
    });
}
