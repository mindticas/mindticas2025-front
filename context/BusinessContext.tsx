'use client';

import { UserProfile } from '@/interfaces/userProfile/userProfile';
import { getUserProfile } from '@/services/userProfileService';
import React, {
    createContext,
    useContext,
    useState,
    ReactNode,
    useEffect,
} from 'react';

interface BusinessInfo {
    name: string;
    address: string;
    phone: string;
    instagram?: string;
}

interface BusinessContextType {
    businessInfo: BusinessInfo;
    setBusinessInfo: (info: BusinessInfo) => void;
    isLoading: boolean;
}

export const BusinessContext = createContext<BusinessContextType | undefined>(
    undefined,
);

export const useBusiness = () => {
    const context = useContext(BusinessContext);
    if (!context) {
        throw new Error('useBusiness must be used within a BusinessProvider');
    }
    return context;
};

export const BusinessProvider: React.FC<{ children: ReactNode }> = ({
    children,
}) => {
    const [businessInfo, setBusinessInfo] = useState<BusinessInfo>({
        name: '',
        address: '',
        phone: '',
        instagram: '',
    });
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const fetchUserProfile = async () => {
        setIsLoading(true);

        try {
            const data = await getUserProfile();
            if (!Array.isArray(data) || data.length === 0) {
                console.error('No se encontrÃ³ perfil del negocio');
                return;
            }
            const profile: UserProfile = data[0];
            setBusinessInfo({
                name: profile.name ?? '',
                address: profile.contactDetails?.address ?? '',
                phone: profile.contactDetails?.phone ?? '',
                instagram: profile.socialLinks?.instagram ?? '',
            });
        } catch (error) {
            console.error('Error al cargar el perfil del usuario', error);
        } finally {
            setIsLoading(false);
        }
    };

    useEffect(() => {
        fetchUserProfile();
    }, []);

    return (
        <BusinessContext.Provider
            value={{ businessInfo, setBusinessInfo, isLoading }}
        >
            {children}
        </BusinessContext.Provider>
    );
};
export type { BusinessInfo };
