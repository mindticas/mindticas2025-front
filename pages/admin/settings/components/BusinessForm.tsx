import {
    Box,
    VStack,
    Text,
    Flex,
    Button,
    Spinner,
    Input,
    Center,
    Skeleton,
} from '@chakra-ui/react';
import React, { useEffect, useState } from 'react';
import { updateUserProfile } from '@/services/userProfileService';
import {
    BusinessFormData,
    hasBusinessFormChanges,
    showBusinessNotification,
    validateBusinessFields,
} from '@/utils/userProfile/userProfileValidation';
import { BusinessInfo } from '@/context/BusinessContext';

interface BusinessFormProps {
    businessInfo: BusinessInfo;
    setBusinessInfo: (info: BusinessInfo) => void;
    isSmallScreen: boolean;
    isLoading: boolean;
}

export default function BusinessForm({
    businessInfo,
    setBusinessInfo,
    isSmallScreen,
    isLoading,
}: BusinessFormProps) {
    const [name, setName] = useState('');
    const [address, setAddress] = useState('');
    const [phone, setPhone] = useState('');
    const [instagram, setInstagram] = useState('');
    const [isSaving, setIsSaving] = useState(false);

    useEffect(() => {
        setName(businessInfo.name);
        setAddress(businessInfo.address);
        setPhone(businessInfo.phone);
        setInstagram(businessInfo.instagram || '');
    }, [businessInfo]);

    const handleSave = async () => {
        setIsSaving(true);
        const trimmedData: BusinessFormData = {
            name: name.trim(),
            contactDetails: {
                address: address.trim(),
                phone: phone.trim(),
            },
            socialLinks: {
                instagram: instagram.trim(),
            },
        };
        // Validation: empty fields
        const validation = validateBusinessFields(trimmedData);
        if (!validation.isValid) {
            setIsSaving(false);
            return showBusinessNotification.validationError(validation.error!);
        }
        // Check for changes in inputs
        if (!hasBusinessFormChanges(businessInfo, trimmedData)) {
            setIsSaving(false);
            return showBusinessNotification.noChanges();
        }
        // If validations pass, save
        try {
            const updated = await updateUserProfile(trimmedData);
            setBusinessInfo({
                name: updated.name || '',
                address: updated.contactDetails?.address || '',
                phone: updated.contactDetails?.phone || '',
                instagram: updated.socialLinks?.instagram || '',
            });
            showBusinessNotification.saveSuccess();
        } catch (error) {
            console.log(error);
            showBusinessNotification.saveError();
        } finally {
            setIsSaving(false);
        }
    };

    return (
        <Box
            width={isSmallScreen ? '100%' : '70%'}
            p={6}
            bg='white'
            borderRadius='lg'
            borderWidth='1px'
            borderColor='gray.200'
            shadow='md'
        >
            {isLoading ? (
                <VStack gap={6} align='stretch'>
                    <Skeleton height='50px' />
                    <Skeleton height='50px' />
                    <Skeleton height='50px' />
                    <Skeleton height='50px' />
                    <Center mt={4}>
                        <Spinner size='xl' color='#1C4ED8' />
                    </Center>
                </VStack>
            ) : (
                <VStack gap={4} align='stretch'>
                    <Box>
                        <Text fontWeight='semibold' mb={2} color='gray.700'>
                            Nombre del negocio
                        </Text>
                        <Input
                            value={name}
                            onChange={(e) => setName(e.target.value)}
                            p={3}
                            borderRadius='md'
                            borderColor='gray.300'
                        />
                    </Box>

                    <Box>
                        <Text fontWeight='semibold' mb={2} color='gray.700'>
                            Dirección
                        </Text>
                        <Input
                            value={address}
                            onChange={(e) => setAddress(e.target.value)}
                            p={3}
                            borderRadius='md'
                            borderColor='gray.300'
                        />
                    </Box>

                    <Box>
                        <Text fontWeight='semibold' mb={2} color='gray.700'>
                            Teléfono
                        </Text>
                        <Input
                            value={phone}
                            onChange={(e) => setPhone(e.target.value)}
                            type='number'
                            p={3}
                            borderRadius='md'
                            borderColor='gray.300'
                        />
                    </Box>

                    <Box>
                        <Text fontWeight='semibold' mb={2} color='gray.700'>
                            Instagram
                        </Text>
                        <Input
                            value={instagram}
                            onChange={(e) => setInstagram(e.target.value)}
                            p={3}
                            borderRadius='md'
                            borderColor='gray.300'
                        />
                    </Box>

                    <Flex justify='center' mt={6}>
                        <Button
                            bgColor='#1C4ED8'
                            color='white'
                            size='md'
                            px={8}
                            py={4}
                            fontWeight='medium'
                            borderRadius='md'
                            onClick={handleSave}
                            _hover={{
                                transform: 'translateY(-2px)',
                                boxShadow: 'md',
                            }}
                            transition='all 0.2s'
                            loading={isSaving}
                        >
                            {isSaving ? (
                                <Flex align='center'>
                                    <Spinner size='sm' color='white' mr={2} />
                                    <Text>Guardando...</Text>
                                </Flex>
                            ) : (
                                'Guardar cambios'
                            )}
                        </Button>
                    </Flex>
                </VStack>
            )}
        </Box>
    );
}
