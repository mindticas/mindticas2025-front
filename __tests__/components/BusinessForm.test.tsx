import React from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import { BusinessInfo } from '@/context/BusinessContext';
import { updateUserProfile } from '@/services/userProfileService';
import {
    validateBusinessFields,
    showBusinessNotification,
} from '@/utils/userProfile/userProfileValidation';
import { render } from '@/utils/render-test';
import BusinessForm from '@/pages/admin/settings/components/BusinessForm';

// More complete mock of the dependencies
jest.mock('@/services/userProfileService', () => ({
    updateUserProfile: jest.fn().mockResolvedValue({
        name: 'Nuevo Nombre',
        contactDetails: { address: 'Nueva Dirección', phone: '987654321' },
        socialLinks: { instagram: '@nuevoinstagram' },
    }),
}));

jest.mock('@/utils/userProfile/userProfileValidation', () => ({
    validateBusinessFields: jest
        .fn()
        .mockReturnValue({ isValid: true, error: null }),
    hasBusinessFormChanges: jest.fn().mockReturnValue(true),
    showBusinessNotification: {
        validationError: jest.fn(),
        noChanges: jest.fn(),
        saveSuccess: jest.fn(),
        saveError: jest.fn(),
    },
}));

describe('BusinessForm', () => {
    const mockBusinessInfo: BusinessInfo = {
        name: 'Mi Negocio',
        address: 'Calle Principal 123',
        phone: '123456789',
        instagram: '@minegocio',
    };

    const mockSetBusinessInfo = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    // Tests de snapshot (se mantienen igual)
    it('must match the snapshot when loading', () => {
        const { asFragment } = render(
            <BusinessForm
                businessInfo={mockBusinessInfo}
                setBusinessInfo={mockSetBusinessInfo}
                isSmallScreen={false}
                isLoading={true}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    describe('Interactions', () => {
        it('You must update the form fields', async () => {
            render(
                <BusinessForm
                    businessInfo={mockBusinessInfo}
                    setBusinessInfo={mockSetBusinessInfo}
                    isSmallScreen={false}
                    isLoading={false}
                />,
            );

            const nameInput = screen.getByDisplayValue('Mi Negocio');
            fireEvent.change(nameInput, { target: { value: 'Nuevo Nombre' } });
            expect(nameInput).toHaveValue('Nuevo Nombre');
        });

        it('must call handleSave when the button is clicked', async () => {
            render(
                <BusinessForm
                    businessInfo={mockBusinessInfo}
                    setBusinessInfo={mockSetBusinessInfo}
                    isSmallScreen={false}
                    isLoading={false}
                />,
            );

            const saveButton = screen.getByText('Guardar cambios');
            fireEvent.click(saveButton);

            await waitFor(() => {
                expect(validateBusinessFields).toHaveBeenCalled();
                expect(updateUserProfile).toHaveBeenCalled();
            });
        });

        it('should display error when validation fails', async () => {
            (validateBusinessFields as jest.Mock).mockReturnValueOnce({
                isValid: false,
                error: 'Validation error',
            });

            render(
                <BusinessForm
                    businessInfo={mockBusinessInfo}
                    setBusinessInfo={mockSetBusinessInfo}
                    isSmallScreen={false}
                    isLoading={false}
                />,
            );

            fireEvent.click(screen.getByText('Guardar cambios'));

            await waitFor(() => {
                expect(
                    showBusinessNotification.validationError,
                ).toHaveBeenCalledWith('Error de validación');
            });
        });
    });
});
