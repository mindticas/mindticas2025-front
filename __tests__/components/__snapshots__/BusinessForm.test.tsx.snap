// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessForm debe coincidir con el snapshot cuando est√° cargando 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-xg2euz"
  >
    <div
      class="chakra-stack css-1l03m6c"
    >
      <div
        class="chakra-skeleton css-k1m867"
      />
      <div
        class="chakra-skeleton css-k1m867"
      />
      <div
        class="chakra-skeleton css-k1m867"
      />
      <div
        class="chakra-skeleton css-k1m867"
      />
      <div
        class="css-sh031d"
      >
        <span
          class="chakra-spinner css-1eysjmw"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
